### DISCUSSION TITLE
Autocompletion for other then standard Python Package does not work with code-server?

### DISCUSSION THREAD

**Discussion by fdervisi** [coder/code-server#4794] [General]
Dear Experts,

I have installed the Python extension and AWS Toolkit, because I need **autocompletion** for the AWS CDK. I build a fresh code-server and only installed those two extensions, but it does not autocomplete the packages for the AWS CDK

It was working on a previous installation, but only when I switch to the python virtual environment, but this time it does not work at all. Do I miss something, or why it is not working?

**Comment by lg-nineoneninethreenine** [comment]
I think we have heard of this before. Can you check Settings (UI) for autocompletion? Or if that doesn't exist, @jsjoeio might have better answers.

**Reply by fdervisi** [reply]
thanks @im-coder-lg , here are the settings for `Auto Complete: Extra Paths` (User)(:

```
{
    "workbench.colorTheme": "Default Light+",
    "json.schemas": [],
    "aws.profile": "profile:default",
    "aws.resources.enabledResources": [
        "AWS::EC2::Host",
        "AWS::EC2::VPC"
    ],
}
```

`Auto Complete: Extra Paths` (Workspace):

```
{
    "python.pythonPath": ".venv/bin/python",
    "python.formatting.provider": "black"
}
```

and `Auto Complete: Extra Paths` (Remote)
```
{
}
```

I can't find `autoComplete.extraPaths`, I tried to add it but I don't know where and where to point it. In my example I always create a virtual environment `python3 -m venv .venv` and source it. My understanding is that then all modules loaded via pip `install -r requirements.txt` should be automatically completed by the extension.

**Reply by fdervisi** [reply]
@im-coder-lg @jsjoeio I can spin up a complete new coder-server on a EC2 at AWS and provide you access via https (caddy)? Then you get an environment to troubleshoot and fix the issue? Please let me know.....

**Reply by jsjoeio** [reply]
@fdervisi that would be awesome! I am not super familiar with the AWS Toolkit and how autocompletion works. The other thing that would be helpful is steps to reproduce this in vanilla VS Code. That way I can see how it's supposed to work.

**Reply by fdervisi** [reply]
@jsjoeio I will send you an e-mail, on your gmail Adress with the credentials and the link, you can do what ever you want, I will destroy it anyhow :)

Here in this [video](https://www.youtube.com/watch?v=I2cXlYYoQqQ) (please jump to **8:40**) you can see how it should work. If you scroll over a **method** of a **class** which you have **imported** you should see all the valid **arguments**:

Example:

```
from constructs import Construct
from aws_cdk import (
    Duration,
    Stack,
    aws_iam as iam,
    aws_sqs as sqs,
    aws_sns as sns,
    aws_sns_subscriptions as subs,
)


class CdkWorkshopStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        queue = sqs.Queue(
            self, "CdkWorkshopQueue",
            visibility_timeout=Duration.seconds(300),
        )

        topic = sns.Topic(
            self, "CdkWorkshopTopic"
        )

        topic.add_subscription(subs.SqsSubscription(queue))
```
If you hover over `sqs` or `Queue`, you will see a **pop up** with all the valid argument / parameters. I don't think it is related to AWS Toolkit, it is just standard python library with documentation for the class/methods. It was working before, but I have refreshed the complete AMI and now it does not work again.

This is on the Ubuntu 20.4 instance:

```
ubuntu@ip-172-31-24-116:~/cdk_workshop$ source /home/ubuntu/cdk_workshop/.venv/bin/activate
(.venv) ubuntu@ip-172-31-24-116:~/cdk_workshop$ pip install -r requirements.txt 
WARNING: The directory '/home/ubuntu/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Collecting aws-cdk-lib==2.8.0
  Downloading aws_cdk_lib-2.8.0-py3-none-any.whl (59.5 MB)
     |████████████████████████████████| 59.5 MB 17.5 MB/s 
Collecting constructs<11.0.0,>=10.0.0
  Downloading constructs-10.0.45-py3-none-any.whl (50 kB)
     |████████████████████████████████| 50 kB 61.0 MB/s 
Collecting jsii<2.0.0,>=1.52.1
  Downloading jsii-1.52.1-py3-none-any.whl (382 kB)
     |████████████████████████████████| 382 kB 67.0 MB/s 
Collecting publication>=0.0.3
  Downloading publication-0.0.3-py2.py3-none-any.whl (7.7 kB)
Collecting typing-extensions<5.0,>=3.7
  Downloading typing_extensions-4.0.1-py3-none-any.whl (22 kB)
Collecting attrs~=21.2
  Downloading attrs-21.4.0-py2.py3-none-any.whl (60 kB)
     |████████████████████████████████| 60 kB 65.7 MB/s 
Collecting cattrs<1.11,>=1.8; python_version >= "3.7"
  Downloading cattrs-1.10.0-py3-none-any.whl (29 kB)
Collecting python-dateutil
  Downloading python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
     |████████████████████████████████| 247 kB 72.3 MB/s 
Collecting six>=1.5
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Installing collected packages: typing-extensions, attrs, cattrs, six, python-dateutil, jsii, publication, constructs, aws-cdk-lib
Successfully installed attrs-21.4.0 aws-cdk-lib-2.8.0 cattrs-1.10.0 constructs-10.0.45 jsii-1.52.1 publication-0.0.3 python-dateutil-2.8.2 six-1.16.0 typing-extensions-4.0.1
(.venv) ubuntu@ip-172-31-24-116:~/cdk_workshop$ 
```

now you just need to open the example code : ~/cdk_workshop/cdk_workshop/cdk_workshop_stack.py and then you can hover over `sqs` or `Queue`

**Reply by jsjoeio** [reply]
Hey @fdervisi! Just wanted to send an update that I received your email and can login. Thanks so much :) I'm a bit strapped on time today but will revisit this first thing tomorrow and see what I can find.

**Reply by fdervisi** [reply]
@jsjoeio thank you very much. I tried now Remote SSH with VS Code on my Mac, and autocompletion works as expected!

So it is most probably a bug in code-server, feel free to trouble shoot on my environment

Thanks for you help!
Cheers
F.

**Reply by lg-nineoneninethreenine** [reply]
Just a sidenote: I think we've seen a bunch of these errors get into Discussions, might be the same error.

**Reply by jsjoeio** [reply]
@jsjoeio Oh wow! That's fantastic. Definitely sounds like a bug in code-server then. We had some messiness with our 4.0.2 release so I might not get to this until next week, but promise I will!

**Reply by lg-nineoneninethreenine** [reply]
Also, the changelog for the latest release seems to have a small summary, I think this wasn't like that, so can anyone investigate? I'll ask @bpmct on Slack if this was intended or happened by mistake.

**Reply by jsjoeio** [reply]
I think that was expected. There wasn't much to share in the release summary 🤔 Well at least based on what @code-asher and I discussed. We could have missed something though

**Reply by jsjoeio** [reply]
Apologies for the delay. I finally found some time to dig in. I'm still having a hard time with this because there are too many factors at play to isolate what the issue could be.

I tried the environment you provided (thank you!) and nothing is popping up on hover

![image](https://user-images.githubusercontent.com/3806031/152032517-249f175f-e88e-40b8-bc78-3c4801d3607f.png)

However, nothing pops up in VS Code either. 

![image](https://user-images.githubusercontent.com/3806031/152032612-d6efec47-8ece-4d15-b638-17741c9d419e.png)

I'm happy to dig more into this for you, but this is what I would need:
- minimal reproducible repro (something I can `git clone` and run an install command). that way i can test in VS Code and code-server
- ideally as few pieces as possible (if this is with any Python package, can you use a different one that's less complex than AWS' Toolkit?)

Then we can figure out what the exact issue is.

Some other things I'll note is we've had issues with Pylance in the past (not sure if that provides autocomplete support or not). We've also had issues with the Python extension so could be that. We won't know though until we get it working as expected in VS Code and then reproduce it in a very very basic environment in code-server.
